panel_segmentation/panel_detection.py

```diff
@@ -541,74 +691,224 @@ def clusterPanels(self, test_mask, fig=False):
+            self.classifyMountingConfiguration(
+                image_file_path=file_name_save_img,
+                acc_cutoff=.65,
+                file_name_save=file_name_save_mount)
+        # Run the azimuth detection pipeline
+        x = imagex.load_img(file_name_save_img,
+                            color_mode='rgb',
+                            target_size=(640, 640))
+        x = np.array(x)
+        # Mask the satellite image
+        res = self.testSingle(x.astype(float), test_mask=None,  model=None)
+        # Use the mask to isolate the panels
+        new_res = self.cropPanels(x, res)
+        plt.imshow(new_res.reshape(640, 640, 3))
+        # # Check azimuth
+        # az = self.detectAzimuth(new_res)
+        # Cluster the solar arrays in the image using connected
+        # components clustering.
+        n, clusters = self.clusterPanels(new_res,
+                                         fig=True)
+        # Calculate the azimuth for each cluster
+        az_list = list()
+        mounting_config_list = list()
+        for ii in np.arange(clusters.shape[0]):
+            az = self.detectAzimuth(clusters[ii][np.newaxis, :])
+            print(az)
+            az_list.append(az)
+            # Overlay the mounting configuration results on top of the
+            # cluster results to classify each cluster's mounting configuration
+            rgb_weights = [0.2989, 0.5870, 0.1140]
+            grayscale_image = np.dot(clusters[ii][np.newaxis, :].reshape(
+                640, 640, 3)[..., :3], rgb_weights)
+            # Loop through all of the mounting configurations and see which one
+            # matches the cluster.
+            pixel_sum = 0
+            mounting_config = None
+            for mounting_config in range(len(labels)):
+                box = boxes[mounting_config]
+                pixel_sum_new = np.sum(
+                    grayscale_image[int(box[1]):int(box[3]),
+                                    int(box[0]):int(box[2])])
+                if pixel_sum_new >= pixel_sum:
+                    pixel_sum = pixel_sum_new
+                    mounting_config = labels[mounting_config]
+                mounting_config_list.append(mounting_config)
+            if len(labels) == 0:
+
+                mounting_config_list.append(["Unknown"]*clusters.shape[0])
+
+        # Update azimuth value if the site is a single-axis tracker
+        # system. This logic handles single axis tracker cases that are
+        # labeled at 90 or 270 (perpendicular to row direction)
+        az_list_updated = list()
+        mounting_config_list_updated = list()
+        for array_number in range(len(az_list)):
+            mount_classification = mounting_config_list[array_number]
+            az = az_list[array_number]
+            if mount_classification == 'ground-single_axis_tracker':
+                if az <= 120:
+                    az = az + 90
+                elif az > 200:
+                    az = az - 90
```

Comment:
Maybe not a big deal, but our friends in the southern hemisphere may prefer an axis azimuth of zero over 180